pfl/privacy/joint_mechanism.py

```diff
@@ -0,0 +1,128 @@
+# Copyright © 2023-2024 Apple Inc.
+'''
+Joint mechanism for combining multiple mechanisms into one.
+'''
+
+from typing import Dict, List, Optional, Set, Tuple
+
+from pfl.metrics import Metrics, StringMetricName
+from pfl.stats import MappedVectorStatistics, TrainingStatistics
+
+from .privacy_mechanism import CentrallyApplicablePrivacyMechanism
+
+
+def check_if_partition(full_set: Set[str], partition: List[Set[str]]):
+    """
+    Checks if each element of full_list appears in exactly one of
+    the sets in partition
+    """
+    partition_union: Set = set()
+    partition_size = 0
+    for s in partition:
+        partition_union = partition_union.union(s)
+        partition_size += len(s)
+
+    return (partition_union == full_set) and (partition_size == len(full_set))
+
+
+class JointMechanism(CentrallyApplicablePrivacyMechanism):
+    """
+    Constructs a new CentrallyApplicablePrivacyMechanism from existing ones.
+    Each existing mechanism is applied to a disjoint subset of the client
+    statistics keys. As such JointMechanism can only be applied to client
+    statistics of type MappedVectorStatistics.
+
+    :param mechanisms_and_keys:
+        Dictionary in which each key is a name of a mechanism and each value
+        is a tuple consisting of the corresponding CentrallyApplicablePrivacyMechanism
+        and a list of keys specifying which portion of the user training statistics that
+        mechanism should be applied to. Note the names of each of the mechanisms must be
+        distinct for the purpose of naming the corresponding Metrics.
+
+    """
+
+    def __init__(
+        self,
+        mechanisms_and_keys: Dict[str,
+                                  Tuple[CentrallyApplicablePrivacyMechanism,
+                                        List[str]]]):
+        if len(set(mechanisms_and_keys.keys())) < len(
+                mechanisms_and_keys.keys()):
+            raise ValueError('Mechanism names must be unique.')
+        self.mechanisms_and_keys = mechanisms_and_keys
+
+    def constrain_sensitivity(
+            self,
+            statistics: TrainingStatistics,
+            name_formatting_fn=lambda n: StringMetricName(n),
+            seed: Optional[int] = None) -> Tuple[TrainingStatistics, Metrics]:
+
+        if not isinstance(statistics, MappedVectorStatistics):
+            raise TypeError(
+                'Statistics must be of type MappedVectorStatistics.')
+
+        if not check_if_partition(
+                set(statistics.keys()),
+            [set(keys) for _, keys in self.mechanisms_and_keys.values()]):
+            raise ValueError(
+                'Mechanism keys do not form a partition of the client statistics keys.'
+            )
+
+        clipped_statistics: MappedVectorStatistics = MappedVectorStatistics()
```

Comment:
`statistics.weight` is lost here. can you make sure it is preserved in the return object and this is discovered in the tests?